glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,305 @@
+            }
+        }});
+        let (tx, rx) = crossbeam::channel::bounded(1 << 10);
+
+        Ok(Self {
+            timer,
+            timer_handler: Some(timer_handler),
+            stall_handler,
+            id: executor_id,
+            terminated,
+            tx,
+            rx,
+        })
+    }
+
+    pub(crate) fn enter_task_queue(
+        &self,
+        queue_name: String,
+        start: Instant,
+        max_expected_runtime: Duration,
+    ) -> nix::Result<StallDetectorGuard<'_>> {
+        // We consider a queue to be stalling the system if it failed to yield in due
+        // time. For a given maximum expected runtime, we allow a margin of error f 10%
+        // (and an absolute minimum of 10ms) after which we record a stacktrace. i.e. a
+        // task queue has should return shortly after `need_preempt()` returns
+        // true or the stall detector triggers. For example::
+        // * If a task queue has a preempt timer of 100ms the the stall detector
+        // triggers if it doesn't yield after running for 110ms.
+        // * If a task queue has a preempt timer of 100ms the the stall detector
+        // triggers if it doesn't yield after running for 15ms.
+
+        let error_margin =
+            Duration::from_millis((max_expected_runtime.as_millis() as f64 * 0.1) as u64)
+                .max(Duration::from_millis(10));
+        StallDetectorGuard::new(
+            self,
+            queue_name,
+            start,
+            max_expected_runtime.saturating_add(error_margin),
+        )
+    }
+
+    pub(crate) fn arm(&self, threshold: Duration) -> nix::Result<()> {
+        self.timer.set(
+            sys::timerfd::Expiration::OneShot(sys::time::TimeSpec::from(threshold)),
+            sys::timerfd::TimerSetTimeFlags::empty(),
+        )
+    }
+
+    pub(crate) fn disarm(&self) -> nix::Result<()> {
+        self.timer.unset()
+    }
+}
+
+impl Drop for StallDetector {
+    fn drop(&mut self) {
+        let timer_handler = self.timer_handler.take().unwrap();
+        self.terminated.store(true, Ordering::Relaxed);
+
+        self.timer
+            .set(
+                sys::timerfd::Expiration::Interval(sys::time::TimeSpec::from(
+                    Duration::from_millis(1),
+                )),
+                sys::timerfd::TimerSetTimeFlags::empty(),
+            )
+            .expect("failed wake the timer for termination");
+
+        let _ = timer_handler.join();
+    }
+}
+
+pub(crate) struct StallDetectorGuard<'detector> {
+    detector: &'detector StallDetector,
+    queue_name: String,
+    start: Instant,
+    threshold: Duration,
+}
+
+impl<'detector> StallDetectorGuard<'detector> {
+    fn new(
+        detector: &'detector StallDetector,
+        queue_name: String,
+        start: Instant,
+        threshold: Duration,
+    ) -> nix::Result<Self> {
+        detector.disarm().unwrap();
+        detector.arm(threshold).unwrap();
```

Comment:
Don't quote me on that because my memory is failing on this, but I am pretty sure we should be able to do a rearm with a single syscall?

disarm + arm, if it does two syscalls, that's a bit much

Reply:
Rearming will implicitly override the old timer. IIRC that's how my original work did it and it worked fine